examples/bazel_env/all_crate_deps_test.sh

```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env bash
+
+# --- begin runfiles.bash initialization v3 ---
+# Copy-pasted from the Bazel Bash runfiles library v3.
+set -uo pipefail; set +e; f=bazel_tools/tools/bash/runfiles/runfiles.bash
+source "${RUNFILES_DIR:-/dev/null}/$f" 2>/dev/null || \
+  source "$(grep -sm1 "^$f " "${RUNFILES_MANIFEST_FILE:-/dev/null}" | cut -f2- -d' ')" 2>/dev/null || \
+  source "$0.runfiles/$f" 2>/dev/null || \
+  source "$(grep -sm1 "^$f " "$0.runfiles_manifest" | cut -f2- -d' ')" 2>/dev/null || \
+  source "$(grep -sm1 "^$f " "$0.exe.runfiles_manifest" | cut -f2- -d' ')" 2>/dev/null || \
+  { echo>&2 "ERROR: cannot find $f"; exit 1; }; f=; set -e
+# --- end runfiles.bash initialization v3 ---
+
+
+set -euo pipefail
+
+# MARK - Functions
+
+fail() {
+  echo >&2 "$@"
+  exit 1
+}
+
+# MARK - Args
+
+if [[ "$#" -ne 1 ]]; then
+  fail "Usage: $0 /path/to/hello_world"
+fi
+HELLO_WORLD="$(rlocation "$1")"
+
+# MARK - Test
```

Comment:
Could you modify this test to invoke all of the tools provided by the upstream_wrapper package, to verify that they all work?

Reply:
Yep, good idea.

Reply:
Maybe separate tests?

Reply:
More like `rustc --version` or a real compilation? (same for the other stuff)

Reply:
I don't have strong opinions - separate tests sound good, but if it's easier to do everything together that's fine too :)

Reply:
> More like rustc --version or a real compilation? (same for the other stuff)

For each, if it's easy to put together a real-world example (i.e. a trivial hello-world binary, with a Cargo project, with a single failing clippy lint or something), that'd be really nice, but if that's really fiddly, just checking that `--version` or `--help` or something for each binary works would be good enough

Reply:
Hmmm... I have been thinking about it. Depending on the context, it is somewhat different what you want.

A) If the user executes something like `bazel run @rules_rust//tools/upstream_wrapper:cargo check`, they probably want it to find the hermetic `rustc`. By default, it will not.

B) Whereas when you create a `bazel_env.bzl`, all the hermetic tools are already in the path and automatically found. The current version works without effort.

---

A1) special handling of sub-commands 

For A), one could bake the dependencies to a couple of things like `rustc` into the wrapper. E.g. it could set the [`RUSTC`](https://doc.rust-lang.org/cargo/reference/environment-variables.html) variable to the hermetic `rustc`.

It might be weird though for a user that if they override `RUSTC` for a reason. So one could maybe check for an existing one? Any ways, it might still be confusing to the user if they have a `rustc` in the path and it doesn't get picked up. 

For `rustc` that might be really an edge case. For external commands, and `cargo clippy` is one, the `cargo-clippy` binary has to be in the `PATH` as far as I know. There is no env variable to specifically tell cargo where to search for this sub command, I think.

A2) put the sysroot bin in the PATH in the wrapper

More logical for me would be to put the sysroot bin directory into the PATH in the wrapper.

To ensure that we don't by accident pick up existing binaries that we have in our sysroot, we'd need to prepend it. Again it might be somewhat confusing in special situations but overall, I think it is the best.